akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolGateway.scala

```diff
@@ -62,13 +62,9 @@ private[http] final class PoolGateway(gatewayRef: ActorRef, val hcps: HostConnec
    * @return a Future completed when the pool has been shutdown.
    */
   def shutdown(): Future[Done] = {
-    shutdownFutureOpt.getOrElse {
-      val shutdownCompletedPromise = Promise[Done]()
+    if (!shutdownCompletedPromise.isCompleted)
```

Comment:
By side effect do you mean passing a message to actor? And what is the "future" on which I should call `foreach`? I thought we should only create futures after sending a message to actor.

Reply:
Yes, the side effect is the message, it could go on line 29 or so, there's always a future connected to a `Promise` (look at the impl for a small surprise ;) ). Something like:
```
shutdownCompletedPromise.foreach { _ => 
  send message to actor
}
```

That will then be triggered when the promise is completed with success, which can only ever happen once.

And then `promise.trySuccess` will not throw if the promise is already completed allowing for multiple calls to `shutdown()` without problems.

Reply:
wouldn't this implementation imply that this shutdownCompletedPromise becomes resolved before we send message to pool actor hence before the actual shutdown would even start?

Reply:
It does, but sending the message does not imply the actor received/processed it yet, so I don't think it makes a very practical difference. It's effectively concurrent with completion of the promise anyway (we could get around that with the Future APIs but I don't think it is worth it given this fact).

If we want to defer completing the future until the pool actually completed shutting down we'll need to introduce new protocol/change how shutdown is done.